src/Microsoft.ML.Data/Model/Onnx/IOnnxContext.cs

```diff
@@ -0,0 +1,109 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using Microsoft.ML.Runtime.Data;
+
+namespace Microsoft.ML.Runtime.Model.Onnx
+{
+    /// <summary>
+    /// A context for defining a ONNX output. The context internally contains the model-in-progress being built. This
+    /// same context object is iteratively given to exportable components via the <see cref="ICanSaveOnnx"/> interface
+    /// and subinterfaces, that attempt to express their operations as ONNX nodes, if they can. At the point that it is
+    /// given to a component, all other components up to that component have already attempted to express themselves in
+    /// this context, with their outputs possibly available in the ONNX graph.
+    /// 
+    /// 
+    /// </summary>
+    public interface IOnnxContext
```

Comment:
Would it be better to make an abstract class instead?  That way we can add more members to this type in the future without having to make an `IOnnxContext2` type.

Reply:
Maybe, maybe. Though in this case, given that the point is that I want *none* of its implementation to be in here, I'm not certain that I see the benefit of a base class. Is there some article I could read that expresses this principle?

Reply:
I think what I'd like to do is lock down this interface/base-class. Is there some way I can do that? Make a base class that no one can easily descend from, except some sort of "approved" implementing assembly? (In this case the onnx project, and also possibly some sort of test assembly.)

Reply:
I'm not sure I get the point about interfaces vs. base classes. If the class would only have abstract methods (as these would), I'm not sure I see how the problem is any better with base classes. And to the extent that you *can* "solve" it by introducing implementations of methods you introduce the fragile base class problem.

`internal` from what I understand works only for restricting access to the same assembly. This certainly doesn't cover this case, since the entire point is that the implementation should be in a *different* assembly. Anything based on signed DLLs, perhaps?

Reply:
> If the class would only have abstract methods (as these would), I'm not sure I see how the problem is any better with base classes.

The point is, you can add members (properties, non-abstract methods) to a class without being a breaking change. New methods can be virtual, and your implementation class override them. And the abstract class can no-op, throw, return a default value, etc.
Once you ship an interface, you can never add new members to it.

Now in this case, if you restrict the class so that ONLY you can inherit from it (like using an `internal` constructor), then you can also add abstract methods to the base class in future versions. Because you know that no one else can inherit from it.

> internal from what I understand works only for restricting access to the same assembly. This certainly doesn't cover this case, since the entire point is that the implementation should be in a different assembly.

You can use `InternalsVisibleTo` to get around this, if you want.